kong/runloop/certificate.lua

```diff
@@ -4,91 +4,139 @@ local http_tls = require "http.tls"
 local openssl_pkey = require "openssl.pkey"
 local openssl_x509 = require "openssl.x509"
 local pl_utils = require "pl.utils"
+local mlcache = require "resty.mlcache"
 
 
 local ngx_log = ngx.log
 local ERR     = ngx.ERR
 local DEBUG   = ngx.DEBUG
+local re_sub  = ngx.re.sub
+local find    = string.find
 
 
 local default_cert_and_key
+local parse_key_and_cert
 
 
 local function log(lvl, ...)
   ngx_log(lvl, "[ssl] ", ...)
 end
 
 
-local function fetch_certificate(sni_name)
-  local row, err = singletons.db.snis:select_by_name(sni_name)
-  if err then
-    return nil, err
+if ngx.config.subsystem == "http" then
+  parse_key_and_cert = function(row)
+    if row == false then
+      return default_cert_and_key
+    end
+
+    -- parse cert and priv key for later usage by ngx.ssl
+
+    local cert, err = ngx_ssl.parse_pem_cert(row.cert)
+    if not cert then
+      return nil, "could not parse PEM certificate: " .. err
+    end
+
+    local key, err = ngx_ssl.parse_pem_priv_key(row.key)
+    if not key then
+      return nil, "could not parse PEM private key: " .. err
+    end
+
+    return {
+      cert = cert,
+      key = key,
+    }
   end
 
-  if not row then
-    log(DEBUG, "no SNI registered for client-provided name: '",
-               sni_name, "'")
-    return true
+else
+  parse_key_and_cert = function(row)
+    if row == false then
+      return default_cert_and_key
+    end
+
+    local ssl_termination_ctx = http_tls.new_server_context()
+    ssl_termination_ctx:setCertificate(openssl_x509.new(row.cert))
+    ssl_termination_ctx:setPrivateKey(openssl_pkey.new(row.key))
+
+    return ssl_termination_ctx
   end
+end
 
-  -- fetch SSL certificate for this sni
 
-  local certificate, err = singletons.db.certificates:select(row.certificate)
-  if err then
-    return nil, err
+local function produce_wild_snis(sni)
+  if type(sni) ~= "string" then
+    error("sni must be a string", 2)
   end
 
-  if not certificate then
-    return nil, "no SSL certificate configured for sni: " .. sni_name
+  local wild_prefix_sni
+  local wild_suffix_sni
+
+  local wild_idx = find(sni, "*", nil, true)
+
+  if wild_idx == 1 then
+    wild_prefix_sni = sni
+
+  elseif not wild_idx then
+    -- *.example.com lookup
+    local wild_sni, n, err = re_sub(sni, [[([^.]+)(\.[^.]+\.\S+)]], "*$2",
+                                    "ajo")
+    if err then
+      log(ERR, "could not create SNI wildcard for SNI lookup: ", err)
+
+    elseif n > 0 then
+      wild_prefix_sni = wild_sni
+    end
   end
```

Comment:
I believe the remaining code in this function could be moved to inside this block, except the `return` statement.
```suggestion

    -- example.* lookup
    wild_sni, n, err = re_sub(sni, [[([^.]+\.)([^.]+)$]], "$1*", "jo")
    if err then
      log(ERR, "could not create SNI wildcard for SNI lookup: ", err)

    elseif n > 0 then
      wild_suffix_sni = wild_sni
    end

  elseif wild_idx == #sni then
    wild_suffix_sni = sni
  end
```

Reply:
Handling those cases separately draws a nicer distinction between leftmost SNI and rightmost SNI. Additionally, we explicitly avoid to nest our code in order to preserve readability (we favor early-exit patterns) and 80 chars limit on column length (as per our contribution and style guidelines). The code was written with both intentents mind.